### PR TITLE
Use specific exception for discovery timeout

### LINKED ISSUE #28149: specific exception for ServiceDiscovery resolve timeout
ServiceDiscovery.resolve takes in a timeout and the implementation is expected to fail the future if the timeout is hit. The API should specify which exception the future will be completed with so it can be recovered from. 

This would be useful in gRPC for re-trying if the look up failed: https://github.com/akka/akka-grpc/blob/master/runtime/src/main/scala/akka/grpc/internal/NettyClientUtils.scala#L43

### PR DESCRIPTION
**Pull request by chbatey**
Unfortunately, as the API is the SPI we can't do this once for all implementations

Fixes #28149

### CODE REVIEW

File: akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala [hunk 1]
```diff
@@ -18,6 +18,14 @@ import akka.util.HashCode
 
 object ServiceDiscovery {
 
+  /**
+   * Future returned by resolve(name, timeout) should be failed with this exception
+   * if the underlying mechanism was unable to resolve the name within the given timeout.
+   *
+   * It is up to each implementation to implement timeouts.
+   */
+  final class DiscoveryTimeoutException() extends Exception
```
Comment (johanandren) [inline akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala]: Wouldn't it be good to allow a message as well, or is the cause of the timeout always obvious?

Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala]: I guess so, I couldn't think of anything apart from saying timed out :)

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala [hunk 2]
```diff
@@ -110,14 +120,17 @@ private[akka] class DnsServiceDiscovery(system: ExtendedActorSystem) extends Ser
...
+        .ask(DnsProtocol.Resolve(srvRequest, mode))(resolveTimeout)
+        .map {
+          case resolved: DnsProtocol.Resolved =>
+            log.debug("{} lookup result: {}", mode, resolved)
+            srvRecordsToResolved(srvRequest, resolved)
+          case resolved =>
+            log.warning("Resolved UNEXPECTED (resolving to Nil): {}", resolved.getClass)
+            Resolved(srvRequest, Nil)
+        }
+        .transform(convertToTimeout)
```
Comment (raboof) [inline akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: I guess we could `recoverWith`, though this is also :+1:

**Review by raboof — APPROVED** [review]

File: akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala (hunk 1 above)
Comment (johanandren) [reply akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala]: Even if we don't have anything to say here, some other discovery implicit could perhaps have several parts/steps that could time out where a desc would be useful (I'm not sure here, just guessing)?

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala (hunk 2 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: :+1: will switch

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala [hunk 3]
```diff
@@ -71,12 +74,17 @@ private[akka] class DnsServiceDiscovery(system: ExtendedActorSystem) extends Ser
...
-  private val dns = if (system.settings.config.getString("akka.io.dns.resolver") == "async-dns") {
-    log.debug("using system resolver as it is set to async-dns")
-    IO(Dns)(system)
-  } else {
-    log.debug("system resolver is not async-dns. Loading isolated resolver")
-    Dns(system).loadAsyncDns("SD-DNS")
+  private val dns = initializeDns()
+
+  // exposed for testing
+  def initializeDns(): ActorRef = {
```
Comment (raboof) [inline akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: `private[dns]`?

**Review by raboof — APPROVED** [review]

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala [hunk 4]
```diff
@@ -299,7 +308,7 @@ abstract class ServiceDiscovery {
    * the passed `resolveTimeout` should never be exceeded, as it signals the application's
    * eagerness to wait for a result for this specific lookup.
    *
-   * The returned future SHOULD be failed once resolveTimeout has passed.
+   * The returned future should be failed once resolveTimeout has passed with a [DiscoveryTimeoutException].
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala]: does it work with single `[  ]`? I think we use double in most places

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala [hunk 5]
```diff
@@ -135,6 +146,11 @@ private[akka] class DnsServiceDiscovery(system: ExtendedActorSystem) extends Ser
     }
   }
 
+  private val convertToTimeout: PartialFunction[Throwable, Future[Resolved]] = {
+    case _: AskTimeoutException =>
+      Future.failed(new DiscoveryTimeoutException("Dns resolve did not respond within resolveTimeout"))
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: can we include the timeout value in the err msg?

**Review by patriknw — COMMENTED** [review]
changed behavior, but I guess fine since we probably didn't document what exception that was expected

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala (hunk 3 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: :+1:

File: akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala (hunk 4 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala]: :+1: doh

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala (hunk 5 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: :+1:

**Review by johanandren — APPROVED** [review]

**Review by raboof — APPROVED** [review]

**Pull request merged**